File: app/src/main/java/com/tao/chinachuclient/RuleActivity.java
package com.tao.chinachuclient;

import android.app.ProgressDialog;
import android.content.Intent;
import android.graphics.Color;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v4.widget.SwipeRefreshLayout.OnRefreshListener;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.ListView;
import android.widget.Toast;

import org.json.JSONException;

import java.io.IOException;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;

import Chinachu4j.Rule;

public class RuleActivity extends AppCompatActivity implements OnRefreshListener, OnItemClickListener{

    private SwipeRefreshLayout swipeRefresh;
    private RuleListAdapter adapter;
    private ApplicationClass appClass;
    private ActionBar actionBar;

    @Override
    protected void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_program);

        appClass = (ApplicationClass)getApplicationContext();

        ListView list = (ListView)findViewById(R.id.programList);
        adapter = new RuleListAdapter(this);
        list.setAdapter(adapter);
        list.setOnItemClickListener(this);

        swipeRefresh = (SwipeRefreshLayout)findViewById(R.id.swipe_refresh);
        swipeRefresh.setColorSchemeColors(Color.parseColor("#2196F3"));
        swipeRefresh.setOnRefreshListener(this);

        actionBar = getSupportActionBar();
        setActionBarTitle(-1);
        actionBar.setDisplayHomeAsUpEnabled(true);

        asyncLoad(false);
    }

    public void asyncLoad(final boolean isRefresh){
        adapter.clear();
        new AsyncTask<Void, Void, Rule[]>(){
            private ProgressDialog progDialog;

            @Override
            protected void onPreExecute(){
                if(!isRefresh){
                    progDialog = new ProgressDialog(RuleActivity.this);
                    progDialog.setMessage(getString(R.string.loading));
                    progDialog.setIndeterminate(false);
                    progDialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
                    progDialog.setCancelable(true);
                    progDialog.show();
                }
            }

            @Override
            protected Rule[] doInBackground(Void... params){
                try{
                    return appClass.getChinachu().getRules();
                }catch(KeyManagementException | NoSuchAlgorithmException | IOException | JSONException e){
                    return null;
                }
            }

            @Override
            protected void onPostExecute(Rule[] result){
                if(!isRefresh)
                    progDialog.dismiss();
                else
                    swipeRefresh.setRefreshing(false);
                if(result == null){
                    Toast.makeText(RuleActivity.this, R.string.error_get_rule, Toast.LENGTH_SHORT).show();
                    return;
                }
                adapter.addAll(result);
                setActionBarTitle(result.length);
            }
        }.execute();
    }

    public void setActionBarTitle(int ruleCount){
        String title = getString(R.string.rule);
        if(ruleCount >= 0){
            title += " (" + ruleCount + ")";
        }
        actionBar.setTitle(title);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item){
        if(item.getItemId() == android.R.id.home){
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onRefresh(){
        asyncLoad(true);
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id){
        Rule rule = (Rule)parent.getItemAtPosition(position);
        Intent i = new Intent(this, RuleDetail.class);
        i.putExtra("position", String.valueOf(position));
        i.putExtra("rule", rule);
        startActivity(i);
    }
}

File: app/src/main/java/com/tao/chinachuclient/RuleActivity.kt
package com.tao.chinachuclient

import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.tao.chinachuclient.databinding.ActivityProgramBinding
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import sugtao4423.library.chinachu4j.Rule

class RuleActivity : AppCompatActivity(), SwipeRefreshLayout.OnRefreshListener, AdapterView.OnItemClickListener {

    private lateinit var binding: ActivityProgramBinding
    private lateinit var adapter: RuleListAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityProgramBinding.inflate(layoutInflater)
        setContentView(binding.root)

        adapter = RuleListAdapter(this)
        binding.programList.let {
            it.adapter = adapter
            it.onItemClickListener = this
        }

        binding.swipeRefresh.let {
            it.setColorSchemeColors(Color.parseColor("#2196F3"))
            it.setOnRefreshListener(this)
        }

        setActionBarTitle(-1)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        asyncLoad(false)
    }

    private fun setActionBarTitle(ruleCount: Int) {
        var title = getString(R.string.rule)
        if (ruleCount >= 0) {
            title += " ($ruleCount)"
        }
        supportActionBar?.title = title
    }

    private fun asyncLoad(isRefresh: Boolean) {
        adapter.clear()
        CoroutineScope(Dispatchers.Main).launch {
            if (!isRefresh) {
                binding.loading.visibility = View.VISIBLE
            }
            val result = withContext(Dispatchers.IO) {
                try {
                    (applicationContext as App).chinachu.getRules()
                } catch (e: Exception) {
                    null
                }
            }
            binding.loading.visibility = View.GONE
            binding.swipeRefresh.isRefreshing = false
            if (result == null) {
                Toast.makeText(this@RuleActivity, R.string.error_get_rule, Toast.LENGTH_SHORT).show()
                return@launch
            }
            adapter.addAll(result)
            setActionBarTitle(result.size)
        }
    }

    override fun onRefresh() {
        asyncLoad(true)
    }

    override fun onItemClick(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
        if (parent == null) {
            return
        }

        startActivity(Intent(this, RuleDetail::class.java).apply {
            putExtra("position", position.toString())
            putExtra("rule", parent.getItemAtPosition(position) as Rule)
        })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

}